index.js

```diff
@@ -139,17 +139,25 @@ chainMethods.forEach(method => {
 })
 
 Chain.prototype.end = function (callback) {
-  if (this._hasInvoked === true) {
+  if (this._hasInvoked === true || this._promise) {
     throw new Error('The dispatch function has already been invoked')
   }
-  doInject(this.dispatch, this.option, callback)
+  if (typeof callback === 'function') {
+    doInject(this.dispatch, this.option, callback)
+  } else {
+    this._promise = doInject(this.dispatch, this.option)
+  }
   this._hasInvoked = true
+  return this
 }
 
 Object.getOwnPropertyNames(Promise.prototype).forEach(method => {
   if (method === 'constructor') return
   Chain.prototype[method] = function (...args) {
     if (!this._promise) {
+      if (this._hasInvoked === true) {
+        throw new Error('The dispatch function has already been invoked')
+      }
```

Comment:
This is not correct. The difference between a promise and end is that one would throw, and one does not, because promises keep the resolved value.

Reply:
Since `end()` would throw if called multiple times, and multiple `then()` returns only one promise, what should we do if one calls `then()` after invoking `end()`? Should we just reject here?

Or these two methods do not effect each other, one can call `then()` after `end()` as normal, v.v.

Reply:
If inside end() you store the promise inject() returns, you can return it here.

Reply:
That's what exactly these method do currently. If you pass no callback function to `end`, it would store the promise, and return the promise in the next `then` call.

Reply:
That's what exactly these method do currently. If you pass no callback function to end, it would store the promise, and return the promise in the next then call.

Is there anything I misunderstand?

Reply:
.end() needs to set the promise as well.

Reply:
Even if there is already a callback?

Reply:
That’s an interesting case, as mixing promises and callbacks would not go well in application. I think we should just set the promise if end() is called without a callback. I would also make the error more explicit to cover the above case.

Reply:
The above case has already been handled. The application would store the promise if `end()` is called without a callback, and then we can access the promise through `then()`.